src/vs/workbench/contrib/terminal/browser/terminalService.ts

```diff
@@ -846,7 +846,7 @@ export class TerminalService extends Disposable implements ITerminalService {
 			instance.onDidFocus(this._onDidChangeActiveInstance.fire, this._onDidChangeActiveInstance),
 			instance.onRequestAddInstanceToGroup(async e => await this._addInstanceToGroup(instance, e))
 		];
-		instance.onDisposed(() => dispose(instanceDisposables));
+		this._register(instance.onDisposed(() => dispose(instanceDisposables)));
```

Comment:
We can't register an instance event against the terminal service as it'll keep stacking up stale references and leaking memory. This probably needs to be tracked in something like `_instanceDisposeListeners: DisposableMap<ITerminalInstance, DisposableStore>` and removed after dispose. Then we can then just register that `DisposableMap` when it's created to clean it up.